includes/traits/llms-trait-earned-engagement-meta-box.php

```diff
@@ -0,0 +1,177 @@
+		'llms_my_certificate' => array(
+			'model'                => 'LLMS_User_Certificate',
+			'user_postmeta_prefix' => '_certificate',
+		),
+	);
+
+	/**
+	 * Add earned engagement fields.
+	 *
+	 * @since [version]
+	 *
+	 * @param array $fields  Array of metabox fields.
+	 * @return array
+	 */
+	protected function add_earned_engagement_fields( $fields = array() ) {
+
+		$post_type = get_post_type();
+		if ( ! array_key_exists( $post_type, $this->allowed_post_types ) ) {
+			return $fields;
+		}
+
+		$student = ! empty( $_GET['sid'] ) ? llms_filter_input( INPUT_GET, 'sid', FILTER_SANITIZE_NUMBER_INT ) : false; // phpcs:ignore
+		$student = empty( $student ) ? ( new $this->allowed_post_types[ $post_type ]['model']( $this->post->ID ) )->get_user_id() : $student;
+
+		if ( empty( $student ) ) {
+			$fields[] = array(
+				'allow_null'      => false,
+				'class'           => 'llms-select2-student',
+				'data_attributes' => array(
+					'allow_clear' => false,
+					'placeholder' => __( 'Select a Student', 'lifterlms' ),
+				),
+				'id'              => $this->prefix . 'student',
+				'label'           => __( 'Select a Student', 'lifterlms' ),
+				'type'            => 'select',
+			);
+		} else {
+			array_unshift(
+				$fields,
+				array(
+					'id'    => $this->prefix . 'student',
+					'type'  => 'hidden',
+					'value' => $student,
+				)
+			);
+		}
+
+		return $fields;
+	}
+
+	/**
+	 * Save a metabox field.
+	 *
+	 * @since [version]
+	 *
+	 * @param int   $post_id WP_Post ID.
+	 * @param array $field   Metabox field array.
+	 * @return boolean
+	 */
+	protected function save_field( $post_id, $field ) {
+
+		if ( $this->prefix . 'student' === $field['id'] && isset( $_POST[ $field['id'] ] ) ) { //phpcs:ignore -- nonce already verified in `LLMS_Admin_Metabox::save`.
+			$this->log_earned_engament( llms_filter_input( INPUT_POST, $field['id'], FILTER_SANITIZE_NUMBER_INT ), $post_id );
+		}
+
+		/**
+		 * Skip saving _llms_student field, only used to award an engagement, it's not a post field.
+		 */
+		if ( isset( $field['id'] ) && $this->prefix . 'student' === $field['id'] ) {
+			return true;
+		}
+
+		parent::save_field( $post_id, $field );
+
+	}
+
+	/**
+	 * Wheter the user has earned the engagement.
+	 *
+	 * @since [version]
+	 *
+	 * @param int $user_id The student's user id.
+	 * @param int $post_id The earned engagement id.
+	 */
+	private function has_user_earned( $user_id, $post_id ) {
+		global $wpdb;
+
+		return (bool) $wpdb->get_var(
+			$wpdb->prepare(
+				"
+				SELECT COUNT( meta_id )
+				FROM {$wpdb->prefix}lifterlms_user_postmeta
+				WHERE user_id=%d
+				AND meta_value=%d
+				",
+				array(
+					$user_id,
+					$post_id,
+				)
+			)
+		);
+	}
+
+	/**
+	 * Wheter the user has earned the engagement.
+	 *
+	 * @since [version]
+	 *
+	 * @param int $user_id The student's user id.
+	 * @param int $post_id The earned engagement id.
+	 * @return void
+	 */
+	private function log_earned_engament( $user_id, $post_id ) {
```

Comment:
On my branch `LLMS_Engagement_Handler::create()` (a private method) stores the user postmeta and triggers this action.

I think the storage of the postmeta and the triggering of the action can be moved into a public method, `LLMS_Engagement_Handler::create_actions( ... )` or something of the sort. You can use that method here and the `create()` method could use it as well.

Reply:
I agree <3

Reply:
I'm going to slip that code then, I guess I'll use a static public method.